### PR TITLE
[WIP] Adding Google T5 model

### PR DESCRIPTION
**Pull request by thomwolf**
Add Google T5 model:
- paper:  "Exploring the Limits of Transfer Learning with a Unified Text-to-Text Transformer" https://arxiv.org/abs/1910.10683
- code: https://github.com/google-research/text-to-text-transfer-transformer

The original model makes heavy use of model and data parallelism to scale the training up to a 11B parameters model. It is based on Mesh TensorFlow (https://github.com/tensorflow/mesh).

We will use a simpler version of model parallelism by simply distributing the layers on several GPUs.

# Workflow for including a model from [README.md](https://github.com/huggingface/transformers/blob/master/templates/adding_a_new_model/README.md)

Here an overview of the general workflow: 

- [x] add model/configuration/tokenization classes
- [x] add conversion scripts
- [x] add tests
- [x] finalize

Let's details what should be done at each step

## Adding model/configuration/tokenization classes

Here is the workflow for adding model/configuration/tokenization classes:

- [x] copy the python files from the present folder to the main folder and rename them, replacing `xxx` with your model name,
- [x] edit the files to replace `XXX` (with various casing) with your model name
- [x] copy-past or create a simple configuration class for your model in the `configuration_...` file
- [x] copy-past or create the code for your model in the `modeling_...` files (PyTorch and TF 2.0)
- [x] copy-past or create a tokenizer class for your model in the `tokenization_...` file

# Adding conversion scripts

Here is the workflow for the conversion scripts:

- [x] copy the conversion script (`convert_...`) from the present folder to the main folder.
- [x] edit this scipt to convert your original checkpoint weights to the current pytorch ones.

# Adding tests:

Here is the workflow for the adding tests:

- [x] copy the python files from the `tests` sub-folder of the present folder to the `tests` subfolder of the main folder and rename them, replacing `xxx` with your model name,
- [x] edit the tests files to replace `XXX` (with various casing) with your model name
- [x] edit the tests code as needed

# Final steps

You can then finish the addition step by adding imports for your classes in the common files:

- [x] add import for all the relevant classes in `__init__.py`
- [x] add your configuration in `configuration_auto.py`
- [x] add your PyTorch and TF 2.0 model respectively in `modeling_auto.py` and `modeling_tf_auto.py`
- [x] add your tokenizer in `tokenization_auto.py`
- [x] add your models and tokenizer to `pipeline.py`
- [x] add a link to your conversion script in the main conversion utility (currently in `__main__` but will be moved to the `commands` subfolder in the near future)
- [x] edit the PyTorch to TF 2.0 conversion script to add your model in the `convert_pytorch_checkpoint_to_tf2.py` file
- [x] add a mention of your model in the doc: `README.md` and the documentation it-self at `docs/source/pretrained_models.rst`.
- [x] upload the pretrained weigths, configurations and vocabulary files.

### CODE REVIEW

**Comment by cfoster0** [comment]
How much work is there remaining on T5? Tried to check out this branch but it looks like the  weights and SentencePiece models aren't uploaded yet.

**Comment by knuser** [comment]
I would like to know if this implementation of T5 is on hold because of priority switching to different tasks or maybe there was found an fundamental obstacle, which causes impossibility to move T5 to transformers library and this pull request will not be finished?

Thanks for great work!

**Comment by thomwolf** [comment]
No fundamental obstacle, I was just stuck on the second place of my TODO list for a while.

This model will need to fine-tuned on a downstream task before usage though because the combination of mesh-tensorflow ops and bfloat16 seems to cause a bigger discrepancy with PyTorch float32 than other architectures.

**Comment by thomwolf** [comment]
Ok merging

**Comment by eelxpeng** [comment]
Good work. Will you be releasing finetuning code soon?

File: transformers/modeling_t5.py [hunk 1]
```diff
@@ -0,0 +1,886 @@
...
+        # Rescale output before projecting on vocab
+        # See https://github.com/tensorflow/mesh/blob/fa19d69eafc9a482aff0b59ddd96b025c0cb207d/mesh_tensorflow/transformer/transformer.py#L586
+        sequence_output = sequence_output * (self.model_dim ** -0.5)
+        lm_logits = self.lm_head(sequence_output)
+
+        decoder_outputs = (lm_logits,) + decoder_outputs[1:]  # Add hidden states and attention if they are here
+        if lm_labels is not None:
+            shift_logits = lm_logits[..., :-1, :].contiguous()
+            shift_labels = lm_labels[..., 1:].contiguous()
+            loss_fct = CrossEntropyLoss(ignore_index=-1)
```
Comment (eelxpeng) [inline transformers/modeling_t5.py]: The loss here is clearly not correct. Should the computation mask out the padding tokens, instead of all steps?

Comment (eelxpeng) [reply transformers/modeling_t5.py]: And in original T5 code, they also seem to be using with label smoothing.
